CheckedPopupMenuItem now fades its icon when changed.

mpcomplete opened the pull request:
BUG=https://github.com/flutter/flutter/issues/2274

mpcomplete commented:
I'm a little sad about how much code is required to add a simple fade animation to the CheckedPopupMenuItem class. Specifically, having to convert it to a StatefulComponent. I'm going to think more about whether this kind of thing can be simplified (maybe allowing simple animations on StatelessComponents).

Hixie commented:
This predates our conversation, right? Or do you still want a review?

Hixie commented:
I'm closing this PR since as I understand it you're still working on it, just so it doesn't appear on the list of open PRs. Let me know if this is wrong. Thanks!

mpcomplete commented:
OK, this is ready for another look. I removed the itemsGetter, and moved the InkWell to the PopupMenuItem's builder so that it has access to the State objects.

HansMuller commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -58,20 +62,31 @@ class PopupMenuItem<T> extends PopupMenuEntry<T> {
+  // Override this to put something else in the menu entry.
+  Widget buildChild() => config.child;
+
+  void onTap(BuildContext context) {
```
I think you can just use the State's context property here; don't need to create a closure around onTap below.

HansMuller commented:
LGTM. This worked out pretty nicely after all.

Hixie commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -41,7 +41,11 @@ class PopupMenuDivider extends PopupMenuEntry<dynamic> {
+}
+
+class _PopupMenuDividerState extends State<PopupMenuDivider> {
+  Widget build(BuildContext context) => new Divider(height: config.height);
```
You can probably get away with having an empty PopupMenuEntry state that everyone uses except `CheckedPopupMenuItem`. Something like:

```dart
class PopupMenuEntryState<T, U extends PopupMenuEntry<T>> extends State<U> {
  Widget build(BuildContext context) => config.build(context);
}
```

Then in PopupMenuEntry:

```dart
  PopupMenuEntryState<T, PopupMenuEntry> createState() => new PopupMenuEntryState<T, PopupMenuEntry>();
```

Then all the build functions can stay on the components, no need for any states except CheckedPopupMenuItem, which would do:

```dart
  _CheckedPopupMenuItemState<T> createState() => new _CheckedPopupMenuItemState<T>();
```

...and _CheckedPopupMenuItemState would be defined as:

```dart
class _CheckedPopupMenuItemState<T> extends PopupMenuEntryState<T, CheckedPopupMenuItem<T>> { ... }
```

Hixie replied:
Well I guess the PopupMenuEntryState would be a bit more complex. It has to introduce the InkWell. The onTap handler would be in there too.

mpcomplete replied:
Yeah, given this, I don't think it's worth trying to be clever with the State object. It would only be used by the Divider.

Hixie commented on packages/flutter/lib/src/material/popup_menu.dart:
```diff
@@ -93,19 +111,52 @@ class CheckedPopupMenuItem<T> extends PopupMenuItem<T> {
+    super.initState();
+    _controller = new AnimationController(duration: _kFadeDuration)
+      ..value = config.checked ? 1.0 : 0.0
+      ..addListener(() => setState(() {}));
```
put a comment in the setState callback saying what state changed (the animation)

Hixie commented:
This is great. LGTM.

Travis anger is an analyzer lint about an unused variable.

mpcomplete committed:
CheckedPopupMenuItem now fades its icon when changed.

BUG=https://github.com/flutter/flutter/issues/2274
